### PR TITLE
[SPARK-36479][SQL][TEST] Improve datetime test coverage in SQL files

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR adds more datetime tests in `date.sql` and `timestamp.sql`, especially for string promotion.

### Why are the changes needed?

improve test coverage

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

N/A

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 1]
```diff
@@ -145,75 +145,20 @@ SELECT INTERVAL '178956970-8' YEAR TO MONTH;
...
-
-select
-  dateval,
-  dateval - interval '2-2' year to month,
-  dateval - interval '-2-2' year to month,
-  dateval + interval '2-2' year to month,
-  dateval + interval '-2-2' year to month,
-  - interval '2-2' year to month + dateval,
-  interval '2-2' year to month + dateval
-from interval_arithmetic;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:163]: moved to `date.sql`

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 2]
```diff
@@ -145,75 +145,20 @@ SELECT INTERVAL '178956970-8' YEAR TO MONTH;
...
-
-select
-  tsval,
-  tsval - interval '2-2' year to month,
-  tsval - interval '-2-2' year to month,
-  tsval + interval '2-2' year to month,
-  tsval + interval '-2-2' year to month,
-  - interval '2-2' year to month + tsval,
-  interval '2-2' year to month + tsval
-from interval_arithmetic;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:173]: moved to `timestamp.sql`

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 3]
```diff
@@ -145,75 +145,20 @@ SELECT INTERVAL '178956970-8' YEAR TO MONTH;
...
-
-select
-  dateval,
-  dateval - interval '99 11:22:33.123456789' day to second,
-  dateval - interval '-99 11:22:33.123456789' day to second,
-  dateval + interval '99 11:22:33.123456789' day to second,
-  dateval + interval '-99 11:22:33.123456789' day to second,
-  -interval '99 11:22:33.123456789' day to second + dateval,
-  interval '99 11:22:33.123456789' day to second + dateval
-from interval_arithmetic;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:190]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 4]
```diff
@@ -145,75 +145,20 @@ SELECT INTERVAL '178956970-8' YEAR TO MONTH;
...
-
-select
-  tsval,
-  tsval - interval '99 11:22:33.123456789' day to second,
-  tsval - interval '-99 11:22:33.123456789' day to second,
-  tsval + interval '99 11:22:33.123456789' day to second,
-  tsval + interval '-99 11:22:33.123456789' day to second,
-  -interval '99 11:22:33.123456789' day to second + tsval,
-  interval '99 11:22:33.123456789' day to second + tsval
-from interval_arithmetic;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:200]: ditto

**Review by yaooqinn — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out [hunk 5]
```diff
@@ -555,28 +589,68 @@ NULL
...
 -- !query schema
 struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
-cannot resolve '(v2.str - DATE '2011-11-11')' due to data type mismatch: argument 1 requires date type, however, 'v2.str' is of string type.; line 1 pos 7
+cannot resolve 'date_add(CAST('1' AS DATE), DATE '2011-11-11')' due to data type mismatch: argument 2 requires (int or smallint or tinyint) type, however, 'DATE '2011-11-11'' is of date type.; line 1 pos 7
+
+
+-- !query
+select date'2011-11-11' + null
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out:628]: This should follow string literal and fail, cc @gengliangwang

File: sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out [hunk 6]
```diff
@@ -555,28 +589,68 @@ NULL
...
+-- !query schema
+struct<date_add(DATE '2011-11-11', NULL):date>
+-- !query output
+NULL
+
+
+-- !query
+select null + date'2011-11-11'
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out:636]: ditto

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 7]
```diff
@@ -1136,117 +1136,157 @@ struct<INTERVAL '-178956970-8' YEAR TO MONTH:interval year to month>
...
+  interval '99 11:22:33.123456789' day to second + '12:12 hour to second',
+  interval '4 22 day to hour' - interval '10' day,
+  interval '2' year + null,
+  null - interval '4 22' day to hour
 -- !query schema
 struct<>
 -- !query output
+org.apache.spark.sql.catalyst.parser.ParseException
+
+Cannot parse the INTERVAL value: 4 22 day to hour(line 5, pos 2)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1163]: I think this is a bug, cc @AngersZhuuuu @MaxGekk

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1163]: > I think this is a bug, cc @AngersZhuuuu @MaxGekk

Got it, will raise a pr and work on this.

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 8]
```diff
@@ -1136,117 +1136,192 @@ struct<INTERVAL '-178956970-8' YEAR TO MONTH:interval year to month>
...
-select
-  tsval,
-  tsval - interval '2-2' year to month,
-  tsval - interval '-2-2' year to month,
-  tsval + interval '2-2' year to month,
-  tsval + interval '-2-2' year to month,
-  - interval '2-2' year to month + tsval,
-  interval '2-2' year to month + tsval
-from interval_arithmetic
+select interval '2' year - '4'
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1186]: I think this should fail.

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 9]
```diff
@@ -1136,117 +1136,192 @@ struct<INTERVAL '-178956970-8' YEAR TO MONTH:interval year to month>
...
-  interval '2-2' year to month - interval '3-3' year to month
-from interval_arithmetic
+select '4 11:11' - interval '4 22:12' day to minute
 -- !query schema
-struct<(INTERVAL '2-2' YEAR TO MONTH + INTERVAL '3-3' YEAR TO MONTH):interval year to month,(INTERVAL '2-2' YEAR TO MONTH - INTERVAL '3-3' YEAR TO MONTH):interval year to month>
+struct<>
 -- !query output
-5-5	-1-1
+java.time.DateTimeException
+Cannot cast 4 11:11 to TimestampType.
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1199]: we should cast string to interval, cc @gengliangwang

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 10]
```diff
@@ -1136,117 +1136,192 @@ struct<INTERVAL '-178956970-8' YEAR TO MONTH:interval year to month>
...
-  interval '99 11:22:33.123456789' day to second + dateval
-from interval_arithmetic
+select '4 12:12:12' + interval '4 22:12' day to minute
 -- !query schema
-struct<dateval:date,dateval - INTERVAL '99 11:22:33.123456' DAY TO SECOND:timestamp,dateval - INTERVAL '-99 11:22:33.123456' DAY TO SECOND:timestamp,dateval + INTERVAL '99 11:22:33.123456' DAY TO SECOND:timestamp,dateval + INTERVAL '-99 11:22:33.123456' DAY TO SECOND:timestamp,dateval + (- INTERVAL '99 11:22:33.123456' DAY TO SECOND):timestamp,dateval + INTERVAL '99 11:22:33.123456' DAY TO SECOND:timestamp>
+struct<>
 -- !query output
-2012-01-01	2011-09-23 12:37:26.876544	2012-04-09 11:22:33.123456	2012-04-09 11:22:33.123456	2011-09-23 12:37:26.876544	2011-09-23 12:37:26.876544	2012-04-09 11:22:33.123456
+java.time.DateTimeException
+Cannot cast 4 12:12:12 to TimestampType.
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1208]: ditto

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 11]
```diff
@@ -1136,117 +1136,213 @@ struct<INTERVAL '-178956970-8' YEAR TO MONTH:interval year to month>
...
+  null - interval '2' year,
+  null + interval '2' hour,
+  null - interval '2' hour
 -- !query schema
-struct<dateval:date,dateval - INTERVAL '2-2' YEAR TO MONTH:date,dateval - INTERVAL '-2-2' YEAR TO MONTH:date,dateval + INTERVAL '2-2' YEAR TO MONTH:date,dateval + INTERVAL '-2-2' YEAR TO MONTH:date,dateval + (- INTERVAL '2-2' YEAR TO MONTH):date,dateval + INTERVAL '2-2' YEAR TO MONTH:date>
+struct<>
 -- !query output
-2012-01-01	2009-11-01	2014-03-01	2014-03-01	2009-11-01	2009-11-01	2014-03-01
+org.apache.spark.sql.AnalysisException
+cannot resolve 'CAST(CAST(NULL AS TIMESTAMP) + INTERVAL '02' HOUR AS VOID)' due to data type mismatch: cannot cast timestamp to void; line 4 pos 2
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:1194]: This is also a bug, cc @gengliangwang

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out [hunk 12]
```diff
@@ -533,136 +533,145 @@ struct<to_timestamp(2019 10:10:10, yyyy hh:mm:ss):timestamp>
...
 -- !query
-select date'2020-01-01' - timestamp'2019-10-06 10:11:12.345678'
+select timestamp'2011-11-11 11:11:11' + null
 -- !query schema
-struct<(DATE '2020-01-01' - TIMESTAMP '2019-10-06 10:11:12.345678'):interval day to second>
+struct<>
 -- !query output
-86 13:48:47.654322000
+org.apache.spark.sql.AnalysisException
+cannot resolve '(TIMESTAMP '2011-11-11 11:11:11' + NULL)' due to data type mismatch: differing types in '(TIMESTAMP '2011-11-11 11:11:11' + NULL)' (timestamp and void).; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:642]: This is a bug, cc @gengliangwang

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out [hunk 13]
```diff
@@ -533,136 +533,145 @@ struct<to_timestamp(2019 10:10:10, yyyy hh:mm:ss):timestamp>
...
 -- !query
-select timestamp'2019-10-06 10:11:12.345678' - date'2020-01-01'
+select null + timestamp'2011-11-11 11:11:11'
 -- !query schema
-struct<(TIMESTAMP '2019-10-06 10:11:12.345678' - DATE '2020-01-01'):interval day to second>
+struct<>
 -- !query output
--86 13:48:47.654322000
+org.apache.spark.sql.AnalysisException
+cannot resolve '(NULL + TIMESTAMP '2011-11-11 11:11:11')' due to data type mismatch: differing types in '(NULL + TIMESTAMP '2011-11-11 11:11:11')' (void and timestamp).; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:651]: ditto

File: sql/core/src/test/resources/sql-tests/results/date.sql.out [hunk 14]
```diff
@@ -498,20 +532,28 @@ struct<date_sub(DATE '2001-10-01', 7):date>
...
+
+
+-- !query
+select date '2001-10-01' - '2001-09-28'
 -- !query schema
 struct<>
 -- !query output
 org.apache.spark.sql.AnalysisException
-cannot resolve 'date_sub(DATE '2001-10-01', CAST('7' AS DOUBLE))' due to data type mismatch: argument 2 requires (int or smallint or tinyint) type, however, 'CAST('7' AS DOUBLE)' is of double type.; line 1 pos 7
+cannot resolve 'date_sub(DATE '2001-10-01', CAST('2001-09-28' AS DOUBLE))' due to data type mismatch: argument 2 requires (int or smallint or tinyint) type, however, 'CAST('2001-09-28' AS DOUBLE)' is of double type.; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/date.sql.out:548]: It's not a breaking change to allow this in non-ansi mode, we can do this later.

**Comment by cloud-fan** [comment]
The GA failure is unrelated, I'm merging this to master/3.2, thanks for the review!

@gengliangwang let's fix the discovered bugs.

**Review by gengliangwang — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/ansi/date.sql.out:628]: The expression `DateAdd` extends the trait `NullIntolerant`. Let's keep the constraints of this trait.

**Pull request closed without merging**